google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreOptions.java

```diff
@@ -216,8 +221,74 @@ public FirestoreOptions build() {
         }
       }
 
+      // Override credentials and channel provider if we are using the emulator.
+      String emulatorHost = System.getenv(FIRESTORE_EMULATOR_SYSTEM_PROPERTY);
+      if (emulatorHost != null) {
+        String hostUrlString = "http://" + emulatorHost;
+
+        // Try creating a host in order to validate that the host name is valid.
+        try {
+          URI.create(hostUrlString);
+        } catch (Exception e) {
+          throw new IllegalArgumentException(
+              "Value: '"
+                  + emulatorHost
+                  + "' for property "
+                  + FIRESTORE_EMULATOR_SYSTEM_PROPERTY
+                  + " is not a valid host",
+              e);
+        }
+
+        // Setting the channel provider to use plaintext turns off SSL.
+        this.setChannelProvider(
+            InstantiatingGrpcChannelProvider.newBuilder()
+                .setEndpoint(emulatorHost)
+                .setChannelConfigurator(
+                    new ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder>() {
+                      @Override
+                      public ManagedChannelBuilder apply(ManagedChannelBuilder input) {
+                        input.usePlaintext();
+                        return input;
+                      }
+                    })
+                .build());
+        this.setCredentials(new FakeCredentials());
+      }
+
       return new FirestoreOptions(this);
     }
+
+    public class FakeCredentials extends Credentials {
+      private Map<String, List<String>> headers;
+
+      public FakeCredentials() {
+        headers = new HashMap<>();
+        headers.put("Authorization", Arrays.asList("Bearer owner"));
+      }
+
+      @Override
+      public String getAuthenticationType() {
+        return "";
```

Comment:
throw an IllegalStateException here, no one should ever call this method

we're returning a nonsense value here, IMO better to immediately fail than let this propagate around